### DISCUSSION TITLE
element_controller.slice_element_by_plane - add function with distance by point?

### DISCUSSION THREAD

**Discussion by roehligas** [cwapi3d/cwapi3dpython#174] [Ideas] [answered]
Hi,

I see throughout my scripts that I end up copying the same function over and over again, and that's the function for calculating the distance requested in the slicing function.

![image](https://github.com/cwapi3d/cwapi3dpython/assets/66385500/1f8df5d7-5f9c-40b2-88d7-e304596682ab)

I'm sure there's a reason for the distance approach, but at least from my user perspective I'm always defining the slicing plane by some point. So to get the distance desired for the slicing one first has to project the slicing point onto a similar angled surface through global zero and calculate the distance there. Couldn't there also be a function that slices by point?

The function below is an attempt to put this all together into one user-friendly function. Ideally project_point_on_plane should also be nested into this function or available somewhere else.

```
def project_point_on_plane(point: cw.point_3d, plane_normal: cw.point_3d, plane_point: cw.point_3d):
    """
        Projects a 3d point onto a 3d plane.
        Returns the closest point on the plane.
    """
    x, y, z = point
    a, b, c = plane_normal
    p, q, r = plane_point

    # Calculate the perpendicular distance from the point to the plane
    distance = (a * (x - p) + b * (y - q) + c * (z - r)) / (a**2 + b**2 + c**2)

    # Calculate the projected point

    x_proj = x - distance * a
    y_proj = y - distance * b
    z_proj = z - distance * c

    projected_point = cw.point_3d(x_proj, y_proj, z_proj)
    
    return projected_point

def split_element_at_point(elem_in:int,normal_vector:cw.point_3d ,cutting_point:cw.point_3d):
    """
    Splits input element with a plane defined by the input vector and the cutting point (offset)
    """
    zp = cw.point_3d(0,0,0)

    projected_cutting_point = project_point_on_plane(zp,normal_vector,cutting_point)
    curr_cut_offset = projected_cutting_point.magnitude()

    # Case: element stretches over global zeropoint, offset direction is flipping
    if angle_between_vectors(normal_vector,projected_cutting_point) > .5:
        curr_cut_offset = -curr_cut_offset
    
    
    new_element = ec.slice_elements_with_plane_and_get_new_elements(elem_in, -normal_vector, -curr_cut_offset)

    return new_element
```

**Comment by Brunner246** [comment, marked as answer]
Maybe the following will help you. You can use the `CwPlane3d` and `CwVector3d` from `cwmath`

```python
import sys
import cadwork
import element_controller as ec

sys.path.append("C:\\...\\.venv\\Lib\\site-packages")

from cwmath.cwplane3d import CwPlane3d
from cwmath.cwvector3d import CwVector3d

if __name__ == '__main__':
    element_ids = ec.get_user_element_ids()
   
    # this could be some element axis (xl, yl, zl)
    plane_normal = CwVector3d(0.000000, -0.500000, 0.866025)
    plane_location = CwVector3d(-42.500000, 1644.450845, 3432.000000)
    plane = CwPlane3d(plane_location, plane_normal)
    distance = plane.distance_to_point(CwVector3d(0.0, 0.0, 0.0))

    for element_id in element_ids:
        print(ec.cut_element_with_plane(element_id,
                                        cadwork.point_3d(*plane_normal),
                                        distance))
```

**Reply by roehligas** [reply]
is there a documentation for cwmath available?

Last time you mentioned it I understood it as something that is going to come?

**Reply by Brunner246** [reply]
At the moment there isnt that much in that project, but the goal is to commit more in future.

https://github.com/cwapi3d/cwmath

**Reply by roehligas** [reply]
Will the documentation later on also be added to this website?

https://docs.cadwork.com/projects/cwapi3dpython/en/latest/

Gonna read myself up in how to get documentation directly from github.

**Reply by Brunner246** [reply]
The goal is that the project is continuously developed (without breaking changes), so you can use it from now.
There is a minimal "documentation" available on https://docs.cadwork.com/projects/cwmath/en/latest/
